pkg/dataplane/server/dataplane.go

```diff
@@ -61,16 +61,31 @@ func (d *Dataplane) GetClusterHost(name string) (string, error) {
 		d.clusters[name].LoadAssignment.GetEndpoints()[0].LbEndpoints[0].GetEndpoint().Hostname, ":")[0], nil
 }
 
-// AddCluster adds a cluster to the map.
+// AddCluster adds/updates a cluster to the map.
 func (d *Dataplane) AddCluster(c *cluster.Cluster) {
 	d.clusters[c.Name] = c
 }
 
+// RemoveCluster adds a cluster to the map.
+func (d *Dataplane) RemoveCluster(name string) {
+	delete(d.clusters, name)
```

Comment:
I just noticed that you need to add a lock to guard d.clusters (in all places where it is accessed).

Reply:
you could potentially encapsulate the map+RWLock (for clusters and listeners, separately) and then add the Update/Delete methods on the new type with internal locking. For Getting the map, you may want to create a slice copy of the names and return that to the caller or handle the clearing internally via a method